### PR TITLE
feature: battle of the interview 

### PR DESCRIPTION
**Pull request by IskandarGhobril**
# Feature: Battle of the Interview

## Description

This PR introduces the **Battle of the Interview** feature, enabling users to challenge each other to mock interviews. The core functionality is implemented, allowing users to initiate and accept challenges. However, the evaluation component of the battles is still under development and will be added in a future update.

A significant part of this PR involved **refactoring existing code** to improve maintainability and reduce duplication. Refactored modules include the **Chat Screen**, **Chat ViewModel**, and **Speaking Practice Module**, making the codebase more modular and reusable.

### Key Features

- **Initiating a Challenge:**
  - Users can navigate to their friends list and select a friend's profile to initiate a battle.
  - Upon initiation, users are redirected to a form to set the context of the battle (e.g., interview type or topic).
  - A battle request is sent to the chosen opponent.

- **Notifications:**
  - The opponent receives a red notification next to the challenger's name.
  - The opponent can click the notification to accept the challenge.
  - **Note:** Rejection functionality is not yet supported.

#### **Main Files**

- `app/src/main/java/com/github/se/orator/ui/battle/`  
  - **`BattleChatScreen.kt`**: Implements the UI for chat functionality during a battle.  
  - **`BattleRequestSentScreen.kt`**: Displays the screen where the user waits for the opponent to accept the battle.  
  - **`BattleScreen.kt`**: Implements the screen where the user challenges their friend by filling out a form.  
  - **`EvaluationScreen.kt`**: Adds the framework for evaluating battles (still under development).  
  - **`WaitingForCompletionScreen.kt`**: Displays a waiting screen while waiting for the opponent to complete the interview.  

- `app/src/main/java/com/github/se/orator/ui/friends/FriendsScreen.kt`  
  - *Updated to include functionality for initiating battles.*

- `app/src/main/java/com/github/se/orator/profile/ProfileScreen.kt`  
  - *Updated to allow users to receive and accept battle requests.*

- `app/src/main/java/com/github/se/orator/model/speechBattle/`  
  - **`BattleRepository.kt`**: Acts as the intermediate layer between the server and the app, managing all battle-related operations.  
  - **`BattleRepositoryFirestore.kt`**: Provides the Firestore implementation of battle-related repository operations.  
  - **`BattleViewModel.kt`**: Manages the state and logic for battles, serving as the bridge between the app's UI and the repository.  
  - **`BattleViewModelFactory.kt`**: Factory for creating instances of the `BattleViewModel`.

- `app/src/main/java/com/github/se/orator/ui/navigation/NavigationActions.kt`  
  - *Added new navigation routes for battle-related screens.*

- `app/src/main/java/com/github/se/orator/network/ChatGPTService.kt`  
  - *Added API integration for ChatGPT-based interaction (to create prompts) during battles and a placeholder for the evaluation.*

- `app/src/main/java/com/github/se/orator/overview/`  
  - **`ChatScreen.kt`**: Refactored to be reusable for the battle chat.  
  - **`SpeakingPracticeModule.kt`**: Refactored logic for speaking practice to make it reusable in battle contexts.  
  - **`SpeakingJobInterviewScreen.kt`**: Adapted to use the refactored code.  
  - **`SpeakingPublicSpeakingScreen.kt`**: Adapted to use the refactored code.  
  - **`SpeakingSalesPitchModule.kt`**: Adapted to use the refactored code.

### Next Steps

- Implementing the evaluation mechanism for battles.
- Adding support for rejecting challenges.
- Adding a cancel mechanism if a user exits the app or navigates back.

### CODE REVIEW

File: app/src/androidTest/java/com/github/se/orator/ui/friends/FriendsUITests.kt [hunk 1]
```diff
@@ -246,11 +264,6 @@ class FriendsUITests {
     verify(mockNavigationActions).navigateTo(eq(Screen.LEADERBOARD))
   }
 
-  /**
-   * Tests that the search functionality works correctly on the View Friends screen. It verifies
-   * that the search field exists and is displayed, can be clicked and text can be inputted. After
-   * inputting text, it checks that the first friend item is displayed and the second is not.
-   */
```
Comment (charbelraffoul) [inline app/src/androidTest/java/com/github/se/orator/ui/friends/FriendsUITests.kt:253]: Hi, the tests you added are great, thanks for your work! I was just wondering if you could revert the changes made to delete comments of the already present tests. I think this would enhance code readability.

File: app/src/main/java/com/github/se/orator/ui/network/ChatGPTService.kt [hunk 2]
```diff
@@ -38,6 +38,11 @@ data class ChatResponse(
     val usage: Usage
 )
 
+enum class SessionType {
+  PRACTICE,
+  BATTLE
```
Comment (charbelraffoul) [inline app/src/main/java/com/github/se/orator/ui/network/ChatGPTService.kt:43]: This Enum class if not being used. I think that it will be useful when implementing the feedback from GPT in the next steps you mentioned. If not, it should be removed.

File: app/src/main/java/com/github/se/orator/ui/offline/RecordingReviewScreen.kt [hunk 3]
```diff
@@ -56,7 +56,7 @@ fun RecordingReviewScreen(
             modifier = Modifier.fillMaxWidth().testTag("Back"),
             verticalAlignment = Alignment.CenterVertically) {
               Icon(
-                  imageVector = Icons.Filled.ArrowBack,
+                  imageVector = Icons.Outlined.ArrowBackIosNew,
```
Comment (charbelraffoul) [inline app/src/main/java/com/github/se/orator/ui/offline/RecordingReviewScreen.kt:59]: I noticed that you changed the arrow back icon on an offline related screen, which seems unrelated to the purpose of this PR. This is a good change as it improves consistency or clarity in the UI. However, while it’s a minor change, it might be better to keep the focus of the PR strictly on its intended scope to maintain clarity in the review process.

File: app/src/main/java/com/github/se/orator/ui/overview/ChatScreen.kt [hunk 4]
```diff
@@ -50,13 +81,6 @@ fun ChatScreen(navigationActions: NavigationActions, chatViewModel: ChatViewMode
     }
   }
 
-  // Initialize and dispose of the conversation when the composable enters or leaves the
-  // composition.
-  //  DisposableEffect(Unit) {
-  //    chatViewModel.initializeConversation()
-  //    onDispose { chatViewModel.endConversation() }
-  //  }
```
Comment (charbelraffoul) [inline app/src/main/java/com/github/se/orator/ui/overview/ChatScreen.kt:58]: It is great that you removed unnecessary comments in order to enhance code readability. Thanks for that!

**Comment by charbelraffoul** [comment]
### Review for PR: *Feature: Battle of the Interview*

---

#### First Impressions:
This is a **great** feature! It brings immense value to the app by seamlessly connecting multiple aspects of the app’s hierarchy—users, friends, and GPT integration—all while introducing a highly engaging and competitive element. This feature not only enhances user interaction but also elevates the app's overall appeal and functionality. Users will appreciate the creativity and effort that went into designing such a well-rounded feature.

The **code quality is excellent**, and it is clear that everything has been meticulously thought through. The modular design and adherence to clean coding principles make the implementation highly maintainable and scalable for future enhancements.

For a PR this massive, spanning over 3,500 lines of code, achieving a **test coverage of over 80%** while keeping **code duplication under 3%** is incredibly impressive. This demonstrates a deep commitment to maintaining high-quality standards even when dealing with complex and extensive changes. Kudos to the effort put into ensuring this balance!

---

### Positive Aspects:

1. **Innovative and Engaging Feature:**
   - The introduction of "Battle of the Interview" is a game-changer. It adds a fun and competitive dynamic to the app, making it more interactive and user-friendly.
   - Integrating battles with GPT-driven content creation further enhances the app's value, giving users a unique and personalized experience.

2. **Thoughtful UI/UX Design:**
   - The notification system for pending battles ensures that users are always aware of challenges, adding a real-time interactive aspect to the app.
   - The screens for initiating challenges, participating in battles, and awaiting results are intuitive and well-structured.

3. **High-Quality Refactoring:**
   - Refactoring the `Chat Screen`, `Chat ViewModel`, and `Speaking Practice Module` to make them more modular and reusable is a great decision. This will significantly simplify future development and maintenance.

4. **Comprehensive Testing:**
   - Achieving **over 80% line coverage** for a feature of this magnitude is an impressive accomplishment. This ensures that the implementation is robust and less prone to bugs.
   - Including more than 2000 lines of tests shows the importance placed on reliability and stability. Thank you for that!

5. **Clean and Modular Code:**
   - The use of dedicated modules and classes for battles (`BattleRepository`, `BattleViewModel`, `BattleScreen`) demonstrates excellent organization and adherence to best practices.
   - The code is easy to read and understand, which will greatly benefit future contributors to the project.

6. **Future-Oriented Design:**
   - Adding a placeholder for the evaluation mechanism and mentioning clear next steps like rejection and cancellation mechanisms shows foresight and preparation for future iterations.

7. **Performance and Scalability:**
   - By keeping the logic modular and ensuring minimal code duplication, the implementation is optimized for scalability as the app grows.

---

### Suggestions for Improvement:

1. **Scope of the PR:**
   - This PR contains more than 3500 added lines, which makes the code harder to review, test and collaborate on. Splitting it in multiple smaller PRs would make the code clearer and easier to collaborate on. 
   - This PR combines feature implementation, refactoring, and unrelated UI enhancements. While all are valuable, they could be split into smaller, more focused PRs for easier review and faster integration.

3. **Commented Code:**
   - There are some commented-out sections in the code. If they are no longer needed, removing them would enhance readability and maintain a cleaner codebase.

4. **Battle Flow Logic:**
   - Consider allowing the initiator to complete the interview immediately and giving the opponent a fixed time (e.g., 24 hours) to complete their part. This would make the experience less dependent on real-time interaction and more flexible for users.

5. **Pending Battles Updates:**
   - Pending battles are currently fetched when the screen is composed, which can cause delays in updates. Adding in-app notifications for real-time updates would enhance user experience and ensure users don't miss challenges. This issue would automatically be solved if we choose to implement the consideration mentioned in suggestion 3.

6. **Live Battles and Friend Status:**
   - If we keep battles live, showing the online status of friends in the list would be a great addition. Users are more likely to challenge online friends, improving engagement and reducing waiting times. This would be done in another PR.

7. **Challenge Flow Intuition:**
   - Clicking on a friend's box to challenge them may not be intuitive for new users. Adding a dedicated "Challenge" button could make the flow more user-friendly.

8. **Cancel Mechanism:**
   - Currently, if a user exits an interview, the opponent is left stuck and cannot go back. Adding a mechanism to handle cancellations or early exits is essential to prevent such scenarios.

9. **Hardcoded Strings:**
   - There are hardcoded strings in the code. While addressing this across the app would require a separate PR, using resource files for new strings should be considered moving forward.

10. **Unrelated UI Enhancements:**
   - Enhancing unrelated UI components, such as the back arrow in offline/RecordingReviewScreen.kt, should ideally be done in a separate PR to keep the focus on the primary feature.

---

### Summary:
This PR is truly exceptional, both in terms of the feature it introduces and the quality of its implementation. The "Battle of the Interview" adds tremendous value to the app by combining user interaction, GPT integration, and competitive elements. The code is clean, modular, and well-tested, showcasing the developer's attention to detail and dedication to quality.

While there are areas for improvement, such as splitting large PRs, handling cancellations, and adding intuitive UI elements, these do not overshadow the incredible effort and thought that went into this PR. With some adjustments and follow-up PRs, this feature will undoubtedly become a standout aspect of the app. I’m really looking forward to seeing the whole feature once the suggestions have been addressed and the next steps are completed!

**Overall: Amazing job! Let me know your thoughts on the suggestions, particularly regarding battle flow, notifications, and live interactions.**

**Review by charbelraffoul — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt [hunk 5]
```diff
@@ -512,4 +513,75 @@ The session is now over. According to the initial instructions, you can now brea
...
+  // $errorBody", e)
+  //                _errorMessage.value = "Failed to send message: ${e.message()}"
+  //            } catch (e: Exception) {
+  //                Log.e("ChatViewModel", "Error sending message: ${e.message}", e)
+  //                _errorMessage.value = "Failed to send message: ${e.message}"
+  //            } finally {
+  //                _isLoading.value = false
+  //            }
+  //        }
+  //    }
```
[the remark this answers is not in the document]
Comment (IskandarGhobril) [reply app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt:549]: These comments came from a pull from the main branch, so I left it because I thought that it might be needed later. I'll make sure to remove it.

File: app/src/main/java/com/github/se/orator/ui/battle/WaitingForCompletionScreen.kt [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+        Column(
+            modifier =
+                Modifier.fillMaxSize()
+                    .padding(innerPadding)
+                    .padding(AppDimensions.paddingMedium)
+                    .wrapContentSize(Alignment.Center)
+                    .testTag("waitingForCompletionScreen"),
+            horizontalAlignment = Alignment.CenterHorizontally) {
+              Text(
+                  text = "You have completed your interview. Waiting for the other user to finish.",
```
[the remark this answers is not in the document]
Comment (IskandarGhobril) [reply app/src/main/java/com/github/se/orator/ui/battle/WaitingForCompletionScreen.kt:67]: I'll make sure to implement this in my next PR. Thanks for the suggestion

File: app/src/main/java/com/github/se/orator/ui/offline/RecordingReviewScreen.kt (hunk 3 above)
Comment (IskandarGhobril) [reply app/src/main/java/com/github/se/orator/ui/offline/RecordingReviewScreen.kt:59]: The code didn't compile due to inconsistencies with the icons after pulling from the main, so I had to change this in order to resolve the issue.

**Review by charbelraffoul — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (IskandarGhobril) [reply app/src/main/java/com/github/se/orator/model/chatGPT/ChatViewModel.kt]: Okay thanks for tracking the issue!

**Comment by charbelraffoul** [comment]
Hi Iskandar, 

I see that all the suggestions have been addressed. Thanks for your efforts! 

I solved a minor merge conflict with main and I will accept your PR now. 

I am looking forward to see this feature once the feedback from GPT is implemented! 

Thanks again for your work!

**Review by charbelraffoul — APPROVED** [review]
